src/main/java/org/jenkinsci/plugins/GithubAuthenticationToken.java

```diff
@@ -289,8 +289,32 @@ public boolean hasOrganizationPermission(String candidateName,
         }
     }
 
-    public boolean hasRepositoryPermission(final String repositoryName) {
-        return myRepositories().contains(repositoryName);
+    public boolean hasRepositoryPermission(final String repositoryName, final Permission permission) {
+        LOGGER.log(Level.FINEST, "Checking for permission: " + permission + " on repo: " + repositoryName + " for user: " + this.userName);
+        boolean isRepoOfMine = myRepositories().contains(repositoryName);
+        if (isRepoOfMine) {
+          return true;
+        }
+        // This is not my repository, nor is it a repository of an organization I belong to.
+        // Check what rights I have on the github repo.
+        GHRepository repository = loadRepository(repositoryName);
+        if (repository == null) {
+          return false;
+        }
+        // let admins do anything
+        if (repository.hasAdminAccess()) {
```

Comment:
As you are caching the whole repository, IIUC you are also storing the permission the requester had at that time. If an admin request the repository the first, all the other users will have admin rights ?

Not sure but it seems to be a security vulnerability, @samrocketman is surely more able to reply on that GitHub aspect.

Reply:
I believe the GitHubAuthenticationToken is per-user - so the cache is effectively per user.

Reply:
In such case you must remove the `static` modifier of the cache :) 

The instance of the object is per user yes but the static variables are shared accross all instances.

Reply:
good catch 👍 

Also, I assume it may be lighter weight to hold some stripped down objects that contain only the data we need to query per-repo, rather than an actual `GHRepository`.

Reply:
It was exactly at the moment I tried to look at what you were using really that I figure out the permissions are stored per user.

A big 👍 if you use a POJO containing only the 3-4 fields you need.